File: R package/R/bits.R
## ID: bits.R, last updated 2020-08-13, F.Osorio

revbits <- function(x)
{ ## linear bit reversal
  if (length(x)) {
    x <- .Fortran("bitreversal",
                  x = as.integer(x),
                  m = as.integer(length(x)))$x
  }
  x
}

File: R package/src/bits.f
c ID: bits.f, last updated 2020-08-13, F.Osorio

cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
      SUBROUTINE bitreversal(c, t)
      INTEGER c(0:*), t
c
c     Elster's linear bit reversal algorithm
c
      INTEGER j, k, n, q
c
      n = 2**t
      k = 1
      c(0) = 0
c
      do q = 0, t - 1
        n = n / 2
        do j = 0, k - 1
          c(k + j) = c(j) + n
        end do
        k = 2 * k
      end do
c
      return
      END
